derotation/simulate/line_scanning_microscope.py

```diff
@@ -0,0 +1,342 @@
+                )
+            ).astype(int)
+            + 1
+        )
+
+        #  round to the nearest even number
+        if self.ps % 2 != 0:
+            self.ps -= 1
+
+        #  final image size depends on the pixel shift
+        self.image_size = line_length - self.ps
+
+    def crop_image(self, image: np.ndarray) -> np.ndarray:
+        """Crop the image to the new size based on the pixel shift."""
+        if self.ps == 0:
+            return image
+        else:
+            return image[
+                # centered in the rows
+                self.ps // 2 : -self.ps // 2,
+                # take the left side of the image
+                : self.image_size,
+            ]
+
+    def rotate_by_line(self) -> np.ndarray:
+        """Simulate the acquisition of a rotated image stack as if for each
+        line acquired, the sample was rotated at a given angle in a given
+        center and plane of rotation.
+
+        Each frame is rotated n_lines_per_frame times, where n_lines_per_frame
+        is the number of lines per frame in the image stack.
+
+        Returns
+        -------
+        np.ndarray
+            The rotated image stack of the same shape as the input image stack,
+            i.e. (num_frames, num_lines_per_frame, num_pixels_per_line).
+        """
+        rotated_image_stack = np.empty(
+            (self.image_stack.shape[0], self.image_size, self.image_size),
+            dtype=np.float64,
+        )
+
+        for i, image in tqdm.tqdm(
+            enumerate(self.image_stack), total=self.image_stack.shape[0]
+        ):
+            is_this_frame_rotating = not np.all(
+                # don't bother if rotation is less than 0.01 degrees
+                np.isclose(self.angles[i], 0, atol=1e-2)
+            )
+            if is_this_frame_rotating:
+                for j, angle in enumerate(self.angles[i]):
+                    if angle == 0:
+                        rotated_image_stack[i][j] = self.crop_image(image)[j]
+                    else:
+                        # rotate the whole image by the angle
+                        rotated_image = self.rotate_sample(image, angle)
+
+                        # if the rotation plane angle is not 0,
+                        # apply the homography
+                        if self.rotation_plane_angle != 0:
+                            rotated_image = self.apply_homography(
+                                rotated_image, "rotation_to_scanning_plane"
+                            )
+                            rotated_image = self.crop_image(rotated_image)
+
+                        # store the rotated image line
+                        rotated_image_stack[i][j] = rotated_image[j]
+            else:
+                rotated_image_stack[i] = self.crop_image(image)
+
+        return rotated_image_stack
+
+    def apply_homography(
+        self, image: np.ndarray, direction: str
+    ) -> np.ndarray:
```

Comment:
Instead of accepting a `str` for direction. Could this method take the matrix itself? Or maybe split into two methods, forward and backward to skip the `if` statement.

Also why is the call to `self.rotate_sample` only necessary when rotating to scanning plane?

Reply:
> Also why is the call to self.rotate_sample only necessary when rotating to scanning plane?
in theory it is not, I'm not using the other direction so I didn't bother to complete it. 😂
BUt given this, since I am only doing from rotation plane to scanning plane, also the if statement is pointless